### PR TITLE
Document Setting up Codefresh CI/CD Pipelines

### PR DESCRIPTION
**Pull request by goruha**
## What
* Document CI/CD steps

### CODE REVIEW

File: content/release-engineering/cicd-process-steps/step_push_image.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+---
+title: "Push docker image"
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_push_image.md]: Proper capitalization

File: content/release-engineering/cicd-process-steps/step_slack_notify.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+---
+title: "Slack notification step"
+description: ""
+weight: 8
+---
+
+After succusfull deploy CI/CD pipeline trigger notification to [Slack](https://slack.com)
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_slack_notify.md]: successful

File: content/release-engineering/cicd-process-steps/step_build_charts.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+---
+title: "Build charts step"
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_build_charts.md]: Capitalization

File: content/release-engineering/cicd-process-steps/step_build_image.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
+---
+title: "Build charts image"
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_build_image.md]: Capitalization. 

Also, I think it should maybe be `Build Chart Image`

File: content/release-engineering/cicd-process-steps/step_deploy.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+---
+title: "Deploy step"
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_deploy.md]: Capitalization

File: content/release-engineering/cicd-process-steps/step_init_variables.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
+---
+title: "Init variables step"
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_init_variables.md]: Capitalization

File: content/release-engineering/cicd-process-steps/step_slack_notify.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+---
+title: "Slack notification step"
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_slack_notify.md]: Capitalization

File: content/release-engineering/cicd-process-steps/step_versioning.md [hunk 8]
```diff
@@ -0,0 +1,26 @@
+---
+title: "Generate versioning step"
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_versioning.md]: `Generate Semantic Versions Step`

File: content/release-engineering/cicd-process-steps/step_versioning.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+---
+
+An Artifacts that are the result of CI/CD build should be stored and versioned
+allow to address them in other CI/CD pipelines builds.
+
+[Semantic versioning standard](https://semver.org/) helps to reduce entropy of
+[dependency hell](https://en.wikipedia.org/wiki/Dependency_hell)
+
+Managing `SemVer` manualy is too complex for developers, so we should automate it.
+We use `Git` as a system of record and generate versions based on it's history.
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/step_versioning.md]: based on it's tag history

**Comment by osterman** [comment]
I think we should rename `cicd-process-steps` to `codefresh-process`

**Comment by osterman** [comment]
Ok. At the very least:

- [ ] drop `-steps` from `cicd-process-steps/`
- [ ] use `-` rather than `_` for filename `step_push_image.md` and drop `step` prefix.

File: content/release-engineering/cicd-process-steps/examples/deploy-with-codefresh.yaml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    title: Deploy app with helmfile
+    image: cloudposse/build-harness:${{BUILD_HARNESS_VERSION}}
+    environment:
+    - AWS_REGION=us-west-2
+    - AWS_ACCESS_KEY_ID=${{AWS_ACCESS_KEY_ID}}          # for chamber
+    - AWS_SECRET_ACCESS_KEY=${{AWS_SECRET_ACCESS_KEY}}  # for chamber
+    - CHART_NAME=app
+    - CHART_REPO_URL=${{STAGING_REPO_ENDPOINT}}
+    - CHART_VERSION=${{SEMVERSION_COMMIT_SHORT}}
+    - EXTERNAL_DNS_ALPHA_KUBERNETES_IO_TARGET=ingress.${{BASE_HOST}}
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/examples/deploy-with-codefresh.yaml]: Where is this used? Not here: https://github.com/cloudposse/geodesic/blob/master/rootfs/conf/kops/helmfile.yaml

File: content/release-engineering/cicd-process-steps/examples/step-versioning-codefresh.yaml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+version: '1.0'
+steps:
+  semver:
+    title: Export semantic version
+    image: cloudposse/build-harness:${{BUILD_HARNESS_VERSION}}
+    commands:
+      - make git/show
```
Comment (osterman) [inline content/release-engineering/cicd-process-steps/examples/step-versioning-codefresh.yaml]: We don't need `git/show`, do we? the `make git/export | see` will output to `stdout`

File: content/glossary/docker-image.md [hunk 12]
```diff
@@ -0,0 +1,9 @@
+---
+title: "Docker Image"
+description: "A [docker image](https://docs.docker.com/engine/reference/commandline/images/) is a self-contained, layered archive containing an application and all of it's OS dependancies and is the artifact of running a `docker build`. The image is what gets stored in a [docker registry](https://docs.docker.com/registry)."
```
Comment (aknysh) [inline content/glossary/docker-image.md]: all of its OS dependancies

File: content/glossary/semver.md [hunk 13]
```diff
@@ -0,0 +1,12 @@
+---
+title: "Semantic Version"
+description: "Semantic versioning (e.g. `1.0.3`) is the most widely adopted scheme for assigning unique version version numbers to software releases."
```
Comment (aknysh) [inline content/glossary/semver.md]: assigning unique version numbers

File: content/glossary/unlimited-staging-environments.md [hunk 14]
```diff
@@ -0,0 +1,8 @@
+---
+title: "Unlimited Staging Environments"
+description: "The ability to run complete, disposable apps on Kubernetes for Staging and Development."
+terms:
+- unlimited staging Environments
```
Comment (aknysh) [inline content/glossary/unlimited-staging-environments.md]: Unlimited Staging Environments

File: content/release-engineering/cicd-process/build-charts.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+A [helm chart](https://docs.helm.sh/developing_charts/#charts) is one of the two artifacts
+that result from the CI/CD build process. The chart defines how to execute the application
+on a Kubernets cluster and what services it depends on.
+
+The [build-harness]({{< relref "release-engineering/build-harness.md" >}}) supports building charts. By default, it will look for a chart in the `charts/` folder located in a project's root folder.
+
+Generally, the process of building a chart looks something like this:
+
+1. Copy the `chart/` folder to a temporary directory
+2. Rewrite `Chart.yaml` and `values.yaml` (using `yq`) to pin the chart version to the docker image to the [semantic version]({{< relref "release-engineering/cicd-process/semantic-versioning.md" >}}).
```
Comment (aknysh) [inline content/release-engineering/cicd-process/build-charts.md]: to pin the chart version to the [semantic version] of the docker image ?

File: content/release-engineering/cicd-process/build-image.md [hunk 16]
```diff
@@ -0,0 +1,23 @@
+---
+title: "Step 3: Build Docker Image"
+description: "Build docker image from `Dockerfile`"
+weight: 3
+tags:
+- cicd
+- codefresh
+- Docker
+- docker image
```
Comment (aknysh) [inline content/release-engineering/cicd-process/build-image.md]: do `tags` allow spaces?

File: content/release-engineering/cicd-process/deploy.md [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+description: "Deploy helm charts to cluster with helmfile and chamber"
+weight: 6
+tags:
+- cicd
+- Codefresh
+- deployment
+---
+
+The deployment step leverages [`helmfile`]({{< relref "tools/helmfile.md" >}}) to define
+[releases](https://docs.helm.sh/using_helm/#three-big-concepts) and orchestrate the deployment them to the cluster with `helm`.
```
Comment (aknysh) [inline content/release-engineering/cicd-process/deploy.md]: and orchestrate the deployment to the cluster

File: content/release-engineering/cicd-process/deploy.md [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+We typically store the `helmfile.yaml` in the application repo under the `config/` directory. If a chart depends on some values, then we store these values in `config/chart.yaml`, which is an `envsubst` formatted template rendered upon deployment.
+
+Here's an example of what a typical helmfile looks like.
```
Comment (aknysh) [inline content/release-engineering/cicd-process/deploy.md]: Here's an example of what a typical `helmfile` looks like: (`:` at the end)

File: content/release-engineering/cicd-process/deploy.md [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+{{% include-code-block title="config/helmfile.yaml" file="release-engineering/cicd-process/examples/app-helmfile.yaml" language="yaml" %}}
+
+Here's an example of what a typical `envsubst` parameterized helm values file looks like.
```
Comment (aknysh) [inline content/release-engineering/cicd-process/deploy.md]: `:` at the end

File: content/release-engineering/cicd-process/build-image.md (hunk 16 above)
Comment (goruha) [reply content/release-engineering/cicd-process/build-image.md]: yes

Comment (goruha) [reply content/release-engineering/cicd-process/build-image.md]: hugo replace space with `-` automaticaly

**Pull request merged**